File: javascript/vehicle-purchase/vehicle-purchase.js
// @ts-check
//
/**
 * Determines whether or not you need a licence to operate a certain kind of vehicle.
 *
 * @param {string} kind
 * @returns {boolean} whether a license is required
 */
export function needsLicense(kind) {
  return ['car', 'truck'].includes(kind);
}

/**
 * Helps choosing between two options by recommending the one that
 * comes first in dictionary order.
 *
 * @param {string} option1
 * @param {string} option2
 * @returns {string} a sentence of advice which option to choose
 */
export function chooseVehicle(option1, option2) {
  const choice = (option1 < option2) ? option1 : option2;
  return choice + ' is clearly the better choice.'
}

/**
 * Calculates an estimate for the price of a used vehicle in the dealership
 * based on the original price and the age of the vehicle.
 *
 * @param {number} originalPrice
 * @param {number} age
 * @returns {number} expected resell price in the dealership
 */
export function calculateResellPrice(originalPrice, age) {
  let discount = 0;

  if (age < 3) {
    discount = 0.8;
  } else if (age <= 10) {
    discount = 0.7;
  } else {
    discount = 0.5;
  }
  
  return originalPrice * discount;
}

File: go/vehicle-purchase/vehicle_purchase.go
package purchase

// NeedsLicense determines whether a license is needed to drive a type of vehicle. Only "car" and "truck" require a license.
func NeedsLicense(kind string) bool {
	return kind == "car" || kind == "truck"
}

// ChooseVehicle recommends a vehicle for selection. It always recommends the vehicle that comes first in lexicographical order.
func ChooseVehicle(option1, option2 string) string {
	var chosen string
	if option1 < option2 {
		chosen = option1
	} else {
		chosen = option2
	}

	return chosen + " is clearly the better choice."
}

// CalculateResellPrice calculates how much a vehicle can resell for at a certain age.
func CalculateResellPrice(originalPrice, age float64) float64 {
	ageFactor := 0.5
	if age < 3 {
		ageFactor = 0.8
	} else if age < 10 {
		ageFactor = 0.7
	}

	return originalPrice * ageFactor
}